### ISSUE TITLE
[Bug]: ROCm build broken for vllm 0.4.0.post1

### ISSUE THREAD

**Issue by vgod-dbx** [vllm-project/vllm#3862]
### Your current environment

docker 26.0.0

### 🐛 Describe the bug

I followed https://docs.vllm.ai/en/latest/getting_started/amd-installation.html#build-from-source-docker-rocm to build a docker image of vllm with ROCm support. 

Building with the default ROCm 6.0 base image worked. However, when I ran the image using `docker run  --network=host --group-add=video --ipc=host --cap-add=SYS_PTRACE --security-opt seccomp=unconfined --shm-size 8G  --device /dev/kfd --device /dev/dri vllm-rocm:v0.4.0.post1 python -m vllm.entrypoints.openai.api_server --model meta-llama/Llama-2-7b-chat-hf`, I got the following error. Seems like a version issue with numpy. 

```
Traceback (most recent call last):
  File "/opt/conda/envs/py_3.9/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/conda/envs/py_3.9/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/vllm-0.4.0.post1+rocm603-py3.9-linux-x86_64.egg/vllm/entrypoints/openai/api_server.py", line 22, in <module>
    from vllm.entrypoints.openai.serving_chat import OpenAIServingChat
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/vllm-0.4.0.post1+rocm603-py3.9-linux-x86_64.egg/vllm/entrypoints/openai/serving_chat.py", line 15, in <module>
    from vllm.model_executor.guided_decoding import (
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/vllm-0.4.0.post1+rocm603-py3.9-linux-x86_64.egg/vllm/model_executor/guided_decoding.py", line 15, in <module>
    from vllm.model_executor.guided_logits_processors import (CFGLogitsProcessor,
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/vllm-0.4.0.post1+rocm603-py3.9-linux-x86_64.egg/vllm/model_executor/guided_logits_processors.py", line 22, in <module>
    from outlines.fsm.fsm import CFGFSM, RegexFSM
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/outlines/__init__.py", line 2, in <module>
    import outlines.generate
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/outlines/generate/__init__.py", line 1, in <module>
    from .api import SequenceGenerator
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/outlines/generate/api.py", line 5, in <module>
    from outlines.fsm.fsm import FSMState
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/outlines/fsm/fsm.py", line 9, in <module>
    from outlines.fsm.regex import create_fsm_index_tokenizer, make_deterministic_fsm
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/outlines/fsm/regex.py", line 5, in <module>
    import numba
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/numba/__init__.py", line 198, in <module>
    _ensure_critical_deps()
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/numba/__init__.py", line 138, in _ensure_critical_deps
    raise ImportError("Numba needs NumPy 1.20 or less")
ImportError: Numba needs NumPy 1.20 or less
```

I tried manually running `pip install numpy==1.20.0`, ran it again, but got another error.
```

Traceback (most recent call last):
  File "/opt/conda/envs/py_3.9/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/conda/envs/py_3.9/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/vllm-0.4.0.post1+rocm603-py3.9-linux-x86_64.egg/vllm/entrypoints/openai/api_server.py", line 22, in <module>
    from vllm.entrypoints.openai.serving_chat import OpenAIServingChat
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/vllm-0.4.0.post1+rocm603-py3.9-linux-x86_64.egg/vllm/entrypoints/openai/serving_chat.py", line 15, in <module>
    from vllm.model_executor.guided_decoding import (
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/vllm-0.4.0.post1+rocm603-py3.9-linux-x86_64.egg/vllm/model_executor/guided_decoding.py", line 15, in <module>
    from vllm.model_executor.guided_logits_processors import (CFGLogitsProcessor,
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/vllm-0.4.0.post1+rocm603-py3.9-linux-x86_64.egg/vllm/model_executor/guided_logits_processors.py", line 22, in <module>
    from outlines.fsm.fsm import CFGFSM, RegexFSM
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/outlines/__init__.py", line 2, in <module>
    import outlines.generate
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/outlines/generate/__init__.py", line 2, in <module>
    from .cfg import cfg
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/outlines/generate/cfg.py", line 5, in <module>
    from outlines.models import OpenAI
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/outlines/models/__init__.py", line 11, in <module>
    from .exllamav2 import ExLlamaV2Model, exl2
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/outlines/models/exllamav2.py", line 5, in <module>
    from .transformers import TransformerTokenizer
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/outlines/models/transformers.py", line 5, in <module>
    from outlines.models.tokenizer import Tokenizer
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/outlines/models/tokenizer.py", line 5, in <module>
    from numpy.typing import NDArray
ImportError: cannot import name 'NDArray' from 'numpy.typing' (/opt/conda/envs/py_3.9/lib/python3.9/site-packages/numpy/typing/__init__.py)
```

Building with ROCm 5.7 as the base image failed at building `numpy`.
```
75.78       numpy/core/src/multiarray/scalartypes.c.src: In function ‘float_arrtype_hash’:                                                                                      
75.78       numpy/core/src/multiarray/scalartypes.c.src:3175:27: error: incompatible type for argument 1 of ‘_Py_HashDouble’                                                    
75.78        3175 |     return _Py_HashDouble((double) PyArrayScalar_VAL(obj, @name@));                                                                                         
75.78       In file included from /opt/conda/envs/py_3.10/include/python3.10/Python.h:77,                                                                                       
75.78                        from numpy/core/src/multiarray/scalartypes.c.src:3:                                                                                                
75.78       /opt/conda/envs/py_3.10/include/python3.10/pyhash.h:10:38: note: expected ‘PyObject *’ {aka ‘struct _object *’} but argument is of type ‘double’                    75.78          10 | PyAPI_FUNC(Py_hash_t) _Py_HashDouble(PyObject *, double);                                                                                                   75.78             |                                      ^~~~~~~~~~                                                                                                             
75.78       numpy/core/src/multiarray/scalartypes.c.src:3175:12: error: too few arguments to function ‘_Py_HashDouble’                                                          
75.78        3175 |     return _Py_HashDouble((double) PyArrayScalar_VAL(obj, @name@));                                                                                         
75.78             |            ^~~~~~~~~~~~~~                                                                                                                                   
75.78       In file included from /opt/conda/envs/py_3.10/include/python3.10/Python.h:77,                                                       
75.78                        from numpy/core/src/multiarray/scalartypes.c.src:3:
75.78       /opt/conda/envs/py_3.10/include/python3.10/pyhash.h:10:23: note: declared here
75.78          10 | PyAPI_FUNC(Py_hash_t) _Py_HashDouble(PyObject *, double);     
75.78             |                       ^~~~~~~~~~~~~~                                                                                                                        
75.78       numpy/core/src/multiarray/scalartypes.c.src: In function ‘cfloat_arrtype_hash’:
75.78       numpy/core/src/multiarray/scalartypes.c.src:3183:31: error: incompatible type for argument 1 of ‘_Py_HashDouble’
75.78        3183 |     hashreal = _Py_HashDouble((double)                                                                                                                      
75.78             |                               ^~~~~~~~ 
75.78             |                               |       
75.78             |                               double                                                                                                                        
75.78        3184 |             PyArrayScalar_VAL(obj, C@name@).real);            
75.78             |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                                                                                             
75.78       In file included from /opt/conda/envs/py_3.10/include/python3.10/Python.h:77,
75.78                        from numpy/core/src/multiarray/scalartypes.c.src:3:
75.78       /opt/conda/envs/py_3.10/include/python3.10/pyhash.h:10:38: note: expected ‘PyObject *’ {aka ‘struct _object *’} but argument is of type ‘double’
75.78          10 | PyAPI_FUNC(Py_hash_t) _Py_HashDouble(PyObject *, double);
75.78             |                                      ^~~~~~~~~~
75.78       numpy/core/src/multiarray/scalartypes.c.src:3183:16: error: too few arguments to function ‘_Py_HashDouble’
75.78        3183 |     hashreal = _Py_HashDouble((double)
75.78             |                ^~~~~~~~~~~~~~
75.78       In file included from /opt/conda/envs/py_3.10/include/python3.10/Python.h:77,
75.78                        from numpy/core/src/multiarray/scalartypes.c.src:3:
75.78       /opt/conda/envs/py_3.10/include/python3.10/pyhash.h:10:23: note: declared here
75.78          10 | PyAPI_FUNC(Py_hash_t) _Py_HashDouble(PyObject *, double);
75.78             |                       ^~~~~~~~~~~~~~
75.78       numpy/core/src/multiarray/scalartypes.c.src:3189:31: error: incompatible type for argument 1 of ‘_Py_HashDouble’
75.78        3189 |     hashimag = _Py_HashDouble((double)
75.78             |                               ^~~~~~~~
75.78             |                               |
75.78             |                               double
75.78        3190 |             PyArrayScalar_VAL(obj, C@name@).imag);
75.78             |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
75.78       In file included from /opt/conda/envs/py_3.10/include/python3.10/Python.h:77,
75.78                        from numpy/core/src/multiarray/scalartypes.c.src:3:
```
(full error message is too long. All errors point to the same `error: too few arguments to function ‘_Py_HashDouble`.)

**Comment by zhuohan123** [comment]
cc @hongxiayang @dllehr-amd. please help cc the right person on the amd side as well :) thanks!

**Comment by hongxiayang** [comment]
@vgod-dbx  etc: I had an image build several days ago with default ROCm 6.0 base image, and it worked fine with your command `python -m vllm.entrypoints.openai.api_server --model `. The commit hash is:
```
f03cc667a09bce92e09365893603f7ec0d87c9f2
```
Can you try that before we can figure out which change could be the source of the problem?

**Comment by vgod-dbx** [comment]
Just tried building a image with your commit hash. Still the same numpy 1.20 issue.
Here are the logs of the build in case it's useful.
 
```
 => [internal] load build definition from Dockerfile.rocm                                                                                                                  0.0s
 => => transferring dockerfile: 3.23kB                                                                                                                                     0.0s
 => [internal] load metadata for docker.io/rocm/pytorch:rocm6.0_ubuntu20.04_py3.9_pytorch_2.1.1                                                                            0.7s
 => [internal] load .dockerignore                                                                                                                                          0.0s
 => => transferring context: 50B                                                                                                                                           0.0s
 => [ 1/17] FROM docker.io/rocm/pytorch:rocm6.0_ubuntu20.04_py3.9_pytorch_2.1.1@sha256:118f9a292165491464f839ba144a712b2a60a15d666f67f25b10d3627da27d8b                    0.0s
 => [internal] load build context                                                                                                                                          0.0s
 => => transferring context: 783.20kB                                                                                                                                      0.0s
 => CACHED [ 2/17] RUN echo "Base image is rocm/pytorch:rocm6.0_ubuntu20.04_py3.9_pytorch_2.1.1"                                                                           0.0s
 => CACHED [ 3/17] RUN echo "FA_GFX_ARCHS is gfx90a;gfx942"                                                                                                                0.0s
 => CACHED [ 4/17] RUN echo "FA_BRANCH is 3d2b6f5"                                                                                                                         0.0s
 => [ 5/17] RUN apt-get update && apt-get install python3 python3-pip -y                                                                                                   4.5s
 => [ 6/17] RUN apt-get update && apt-get install -y     curl     ca-certificates     sudo     git     bzip2     libx11-6     build-essential     wget     unzip     nv  161.7s
 => [ 7/17] WORKDIR /app                                                                                                                                                   0.0s 
 => [ 8/17] RUN python3 -m pip install --upgrade pip                                                                                                                       2.7s 
 => [ 9/17] RUN python3 -m pip install --no-cache-dir fastapi ninja tokenizers pandas                                                                                     10.2s 
 => [10/17] RUN if [ "1" = "1" ]; then     mkdir libs     && cd libs     && git clone https://github.com/ROCm/flash-attention.git     && cd flash-attention     && git   662.8s 
 => [11/17] RUN if [ "rocm/pytorch:rocm6.0_ubuntu20.04_py3.9_pytorch_2.1.1" = "rocm/pytorch:rocm6.0_ubuntu20.04_py3.9_pytorch_2.1.1" ]; then     rm -rf /opt/conda/envs/p  0.1s 
 => [12/17] COPY ./ /app/vllm                                                                                                                                              0.1s 
 => [13/17] RUN python3 -m pip install --upgrade pip                                                                                                                       1.6s 
 => [14/17] RUN python3 -m pip install xformers==0.0.23 --no-deps                                                                                                         13.1s 
 => [15/17] RUN cd /app     && cd vllm     && pip install -U -r requirements-rocm.txt     && if [ "1" = "1" ]; then        bash patch_xformers.rocm.sh; fi     && patch  246.5s 
 => [16/17] RUN python3 -m pip install --upgrade pip                                                                                                                       1.9s 
 => [17/17] RUN python3 -m pip install --no-cache-dir ray[all]==2.9.3                                                                                                     18.0s 
 => exporting to image                                                                                                                                                     8.2s 
 => => exporting layers                                                                                                                                                    8.2s 
 => => writing image sha256:9dd0bf284c8aec713ebc25bd88047903b2b9635f49220a6b0cc6ccec834aa16f                                                                               0.0s 
 => => naming to docker.io/library/vllm-rocm:sha-f03cc66                                                                                                                   0.0s 
```

Here are the list of pip libraries I got using `pip freeze` in the container.

```
absl-py==2.0.0
aiohttp==3.9.1
aiohttp-cors==0.7.0
aiorwlock==1.4.0
aiosignal==1.3.1
annotated-types==0.6.0
anyio==4.3.0
apex @ file:///var/lib/jenkins/apex
asgiref==3.7.2
astunparse==1.6.3
async-timeout==4.0.3
attrs==23.1.0
audioread==3.0.1
blessed==1.20.0
boto3==1.19.12
botocore==1.22.12
cachetools==5.3.2
certifi==2023.11.17
cffi==1.16.0
charset-normalizer==3.3.2
click==8.1.7
cloudpickle==3.0.0
cmake==3.29.0.1
colorama==0.4.6
colorful==0.5.6
coremltools==5.0b5
Cython==3.0.6
decorator==5.1.1
Deprecated==1.2.14
diskcache==5.6.3
distlib==0.3.8
Django==4.2.8
dm-tree==0.1.8
einops==0.7.0
exceptiongroup==1.2.0
execnet==2.0.2
expecttest==0.1.6
Farama-Notifications==0.0.4
fastapi==0.108.0
filelock==3.13.3
flash-attn==2.0.4
flatbuffers==2.0
frozenlist==1.4.1
fsspec==2024.3.1
future==0.18.3
geojson==2.5.0
ghstack==0.7.1
google-api-core==2.18.0
google-auth==2.25.2
google-auth-oauthlib==1.0.0
googleapis-common-protos==1.63.0
gpustat==1.1.1
grpcio==1.60.0
gymnasium==0.28.1
h11==0.14.0
httptools==0.6.1
huggingface-hub==0.22.2
hypothesis==5.35.1
idna==3.6
image==1.5.33
imageio==2.33.1
importlib-metadata==7.0.0
iniconfig==2.0.0
interegular==0.3.3
jax-jumpy==1.0.0
Jinja2==3.1.2
jmespath==0.10.0
joblib==1.3.2
jsonschema==4.21.1
jsonschema-specifications==2023.12.1
junitparser==2.1.1
lark==1.1.9
lazy_loader==0.3
librosa==0.10.1
lintrunner==0.10.7
llvmlite==0.37.0
lxml==4.9.3
lz4==4.3.3
Markdown==3.3.4
markdown-it-py==3.0.0
MarkupSafe==2.1.3
mdurl==0.1.2
mkl-fft==1.3.1
mkl-random==1.2.2
mkl-service==2.4.0
mpmath==1.3.0
msgpack==1.0.7
multidict==6.0.4
mypy==1.4.1
mypy-extensions==1.0.0
nest-asyncio==1.6.0
networkx==2.8.8
ninja==1.11.1.1
numba==0.54.1
numpy==1.22.4
nvidia-ml-py==12.535.133
oauthlib==3.2.2
opencensus==0.11.4
opencensus-context==0.1.3
opentelemetry-api==1.24.0
opentelemetry-exporter-otlp==1.24.0
opentelemetry-exporter-otlp-proto-common==1.24.0
opentelemetry-exporter-otlp-proto-grpc==1.24.0
opentelemetry-exporter-otlp-proto-http==1.24.0
opentelemetry-proto==1.24.0
opentelemetry-sdk==1.24.0
opentelemetry-semantic-conventions==0.45b0
opt-einsum==3.3.0
optionloop==1.0.7
outlines==0.0.34
packaging==23.2
pandas==2.2.1
Pillow==9.5.0
platformdirs==4.1.0
pluggy==1.3.0
pooch==1.8.0
prometheus_client==0.20.0
proto-plus==1.23.0
protobuf==3.20.2
psutil==5.9.8
py-cpuinfo==9.0.0
py-spy==0.3.14
pyarrow==15.0.2
pyasn1==0.5.1
pyasn1-modules==0.3.0
pycparser==2.21
pydantic==2.6.4
pydantic_core==2.16.3
Pygments==2.15.0
pytest==7.3.2
pytest-cpp==2.3.0
pytest-flakefinder==1.1.0
pytest-rerunfailures==13.0
pytest-shard==0.1.2
pytest-xdist==3.3.1
python-dateutil==2.8.2
python-dotenv==1.0.1
pytz==2024.1
PyWavelets==1.4.1
PyYAML @ file:///croot/pyyaml_1698096049011/work
ray==2.9.3
ray-cpp==2.9.3
referencing==0.34.0
regex==2023.12.25
requests==2.31.0
requests-oauthlib==1.3.1
rich==13.7.1
rockset==1.0.3
rpds-py==0.18.0
rsa==4.9
s3transfer==0.5.2
safetensors==0.4.2
scikit-image==0.19.3
scikit-learn==1.3.2
scipy==1.6.3
sentencepiece==0.2.0
shellingham==1.5.4
six @ file:///tmp/build/80754af9/six_1644875935023/work
smart-open==7.0.4
sniffio==1.3.1
sortedcontainers==2.4.0
soundfile==0.12.1
soxr==0.3.7
sqlparse==0.4.4
starlette==0.32.0.post1
sympy==1.12
tb-nightly==2.13.0a20230426
tensorboard==2.13.0
tensorboard-data-server==0.7.2
tensorboardX==2.6.2.2
threadpoolctl==3.2.0
tifffile==2023.12.9
tiktoken==0.6.0
tokenizers==0.15.2
tomli==2.0.1
torch @ file:///var/lib/jenkins/pytorch/dist/torch-2.1.1%2Bgit011de5c-cp39-cp39-linux_x86_64.whl#sha256=f08eedebce0dfc36dc38c0152851868642d7262ba270b5543736ede8c129bd87
torchvision==0.16.1+fdea156
tqdm==4.66.1
transformers==4.39.3
triton @ git+https://github.com/ROCmSoftwarePlatform/triton@066e7deb0b8394c73ca67a4d31cde708018b1f13#subdirectory=python
typer==0.12.1
typing_extensions==4.11.0
tzdata==2024.1
unittest-xml-reporting==3.2.0
urllib3==1.26.18
uvicorn==0.29.0
uvloop==0.19.0
virtualenv==20.25.1
vllm==0.4.0+rocm603
watchfiles==0.21.0
wcwidth==0.2.13
websockets==12.0
Werkzeug==3.0.1
wrapt==1.16.0
xdoctest==1.1.0
xformers==0.0.23
yarl==1.9.4
z3-solver==4.12.2.0
zipp==3.17.0
```

**Comment by mawong-amd** [comment]
Hi @vgod-dbx, I replaced line 80 in the Dockerfile `RUN python3 -m pip install --upgrade pip` with `RUN python3 -m pip install --upgrade pip numba` and I seem to be able to run the server (using your command exactly) using that built Docker.

**Comment by vgod-dbx** [comment]
@mawong-amd Thanks for the quick response I can confirm that the fix works for ROCm 6.0.

**Comment by linchen111** [comment]
> @mawong-amd Thanks for the quick response. I can confirm that the fix works for ROCm 6.0.

I failed in MI250x , is it possible for you to share your image? Thanks!
